assets/js/googlesitekit/widgets/datastore/widgets.js

```diff
@@ -0,0 +1,260 @@
+ * we can't store React components in our data store.
+ *
+ * @private
+ * @since n.e.x.t
+ */
+export const WidgetComponents = {};
+
+const ASSIGN_WIDGET = 'ASSIGN_WIDGET';
+const REGISTER_WIDGET = 'REGISTER_WIDGET';
+const SET_WIDGET_COMPONENT_KEY = 'SET_WIDGET_COMPONENT_KEY';
+
+export const INITIAL_STATE = {
+	registryKey: undefined,
+	widgets: {},
+};
+
+export const actions = {
+	/**
+	 * Assigns an existing widget (by slug) to a widget area(s).
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}         slug      Widget slug.
+	 * @param {(string|Array)} areaSlugs Widget Area slug(s).
+	 * @return {Object} Redux-style action.
+	 */
+	assignWidget( slug, areaSlugs ) {
+		return {
+			payload: {
+				slug,
+				areaSlugs: ( typeof areaSlugs === 'string' ) ? [ areaSlugs ] : areaSlugs,
+			},
+			type: ASSIGN_WIDGET,
+		};
+	},
+
+	/**
+	 * Register a widget with a given slug and settings.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param {string}          slug               Widget's slug.
+	 * @param {Object}          settings           Widget's settings.
+	 * @param {React.Component} settings.component React component used to display the contents of this widget.
+	 * @param {number}          settings.priority  Optional. Widget's priority for ordering (lower number is higher priority, like WordPress hooks). Default is: 10.
+	 * @param {string}          settings.width     Optional. Widget's maximum width to occupy. Default is: "quarter". One of: "quarter", "half", "full".
+	 * @return {Object} Redux-style action.
+	 */
+	*registerWidget( slug, {
+		component,
+		priority = 10,
+		width = 'quarter',
+	} = {} ) {
+		invariant( component, 'component is required to register a widget.' );
+		invariant( WIDGET_WIDTHS.includes( width ), `Widget width should be one of: ${ WIDGET_WIDTHS.join( ', ' ) } but "${ width }" was provided.` );
+
+		const registry = yield commonActions.getRegistry();
+		let registryKey = yield registry.select( STORE_NAME ).getWidgetRegistryKey();
+
+		if ( registryKey === undefined ) {
+			registryKey = Object.keys( WidgetComponents ).length + 1;
+			yield {
+				payload: { registryKey },
+				type: SET_WIDGET_COMPONENT_KEY,
+			};
+		}
+
+		// We do this assignment in the action rather than the reducer because we can't send a
+		// payload that includes a React component to the reducer; we'll get an error about
+		// payloads needing to be plain objects.
+		if ( WidgetComponents[ registryKey ] === undefined ) {
+			WidgetComponents[ registryKey ] = {};
+		}
+		if ( WidgetComponents[ registryKey ][ slug ] === undefined ) {
+			WidgetComponents[ registryKey ][ slug ] = component;
+		}
+
+		yield {
+			payload: { slug, settings: { priority, width } },
+			type: REGISTER_WIDGET,
+		};
+
+		return {};
+	},
+};
+
+export const controls = {};
+
+export const reducer = ( state, { type, payload } ) => {
+	switch ( type ) {
+		case ASSIGN_WIDGET: {
+			const { slug, areaSlugs } = payload;
+
+			const { areaWidgets } = state;
```

Comment:
When I asked to remove that from `areas.js`, I didn't think about this being the storage for the widget _assignments_. Now that this is not in any initial state, it makes some tests fail. It should be added to the `INITIAL_STATE` of this file, but under a more clear name (see above), e.g. `areaAssignments` or `areaWidgetAssignments`.

Reply:
I was testing this locally in VSCode and must've removed that bit last and just not saw the test failure, my bad! 😅